Add customizable banner component

choldgraf opened the pull request:
Adds support for customizable announcement banners and sidebar footers in the book theme. This would let us display a "banner announcement" similar to JB1. I had to do some guessing about the best way to do this but I basically tried to follow the same patterns we followed in #565 

I also had a hard time getting a build to work with the XIcon, and had to make it TwitterIcon to avoid an error, it's not related to this enhancement but left it in there in case that's a fix for something?

Major changes:

- Add a dismissible Banner component with localStorage persistence so it only re-shows if the content changes
- Add an ad-hoc hash utility for generating stable banner IDs from content
- Fix netlify builds due to a new python version clash w/ Altair
- Added some docs about all of this so that the behavior is documented here (we should add to the myst/jb docs too)

Refs

- closes jupyter-book/myst-theme#509

agoose77 commented on themes/book/app/routes/$.tsx:
```diff
@@ -90,14 +91,16 @@ function ArticlePageAndNavigationInternal({
+  const projectParts = siteManifest?.projects?.[0]?.parts;
   return (
     <>
+      {projectParts?.banner && <Banner content={projectParts.banner.mdast} />}
```
```suggestion
      <TabStateProvider>
      {projectParts?.banner && <Banner content={projectParts.banner.mdast} />}
      </TabStateProvider>
```

choldgraf commented:
Thanks for adding the tabstate stuff, I think that's what you meant about "context"? I added a short docs note on that in the contributing guide. I also removed the primary sidebar footer stuff.

choldgraf commented on themes/book/app/components/Banner.tsx:
```diff
@@ -0,0 +1,74 @@
+  // Check dismissal state on client side only
+  // If the banner content changes, the ID will be different and it'll show again
+  useEffect(() => {
+    const dismissed = localStorage.getItem(`dismissed-banner-${bannerId}`) === 'true';
```
IMO that makes a lot of sense, +1 from me.

choldgraf commented on themes/book/app/utils/hash.ts:
```diff
@@ -0,0 +1,16 @@
+ * Generates a simple hash from a string for creating stable IDs
+ */
+export function hashString(content: string): string {
+  let hash = 0;
```
can you show me how to do that? I felt like I must be missing some kind of internal utility function but didn't know enough to find it

stefanv replied:
From https://mojoauth.com/hashing/sha-256-in-typescript/:

```ts
import { createHash } from 'crypto';
/**
 * Function to generate a SHA-256 hash of a given input.
 * @param input - The input string to hash.
 * @returns The SHA-256 hash as a hexadecimal string.
 */
function generateSHA256(input: string): string {
    return createHash('sha256').update(input).digest('hex');
}
// Example usage
const input = "Hello, world!";
const hash = generateSHA256(input);
console.log(`SHA-256 Hash: ${hash}`);
```

agoose77 replied:
I think the example that comes to mind is in the `myst-execute` package, in `mystmd`.

choldgraf replied:
ok will try using this

choldgraf commented on themes/book/app/components/Banner.tsx:
```diff
@@ -0,0 +1,74 @@
+import classNames from 'classnames';
+import type { GenericParent } from 'myst-common';
+import { hashString } from '~/utils/hash';
+import { XMarkIcon } from '@heroicons/react/24/solid';
```
well one is from `scienceicons` and this is from `heroicons`, so that could explain the difference right?

choldgraf commented on themes/book/app/routes/$.tsx:
```diff
@@ -90,9 +91,13 @@ function ArticlePageAndNavigationInternal({
   return (
     <>
+      <TabStateProvider>
+      {projectParts?.banner && <Banner content={projectParts.banner.mdast} />}
```
my understanding was that we only use tailwind classes for styling, but I'm definitely +1 on adding myst-specific classnames (`myst--banner`?) if folks are happy w/ that

stefanv commented on docs/ui.md:
```diff
@@ -0,0 +1,44 @@
+- Users can dismiss it by clicking the X button
+- Dismissal persists in the browser (using `localStorage`)
+- Will expand vertically if extra content is in it
+- If you change the banner content, it will reappear for all users
```
I think if this remains part of the myst-theme, docs, which users do not primarily use, then it's perfectly fine.

choldgraf replied:
cool - let's leave it, and see how it feels in practice. I tried to clarify the purpose of these docs in the rEADME

choldgraf commented on docs/ui.md:
```diff
@@ -0,0 +1,44 @@
+# User Interface Components
```
ok I'll add a little note in the contributing docs

choldgraf commented:
OK I *think* that I made all the suggestions above!

agoose77 commented on docs/developer.md:
```diff
@@ -39,9 +39,30 @@ make build-docs
+When creating components that render MyST content (using `<MyST ast={content} />`), wrap them in a `TabStateProvider` to ensure proper React context is available.
+This ensures that interactivity works properly, and doesn't negatively impact _other_ interactive content on a page.
+
+See [`themes/book/app/routes/$.tsx#L98-L121`](https://github.com/jupyter-book/myst-theme/blob/f758f671a883bbdc2e9950ed2da9c0b99fb96720/themes/book/app/routes/$.tsx#L98-L121) for examples of how the Banner and Footer components are wrapped in `TabStateProvider`.
```
We will need to add more here before we merge. It's not just `TabStateProvider` -- there are several other providers that sometimes need to be brought in.

Let me try and rework this


```suggestion
### Using Providers for MyST content

When using components that render MyST content (using `<MyST ast={content} />`), they must be wrapped wrap them in various providers e.g. `TabStateProvider` to ensure proper React context is available.
These providers facilitate an alternative to prop drilling, the name used to describe passing props down through several React components. They are also used to isolate state between different subtrees, e.g. `TabStateProvider` defines a synchronised tab context.

See [`themes/book/app/routes/$.tsx#L98-L121`](https://github.com/jupyter-book/myst-theme/blob/f758f671a883bbdc2e9950ed2da9c0b99fb96720/themes/book/app/routes/$.tsx#L98-L121) for examples of how the Banner and Footer components are wrapped in `TabStateProvider`.
```

agoose77 commented:
Thanks for adding this @choldgraf, and for reviewing @stefanv! I think we can merge!
